### ISSUE TITLE
[FR] Add option to stash files while keeping local changes.

### ISSUE THREAD

**Issue by vovodroid** [gitextensions/gitextensions#10905]
### Feature description

Sometimes it's very handy to use stash as temporary backup for work in progress without discarding files. Sure, it's always possible to stash and next apply it, but it causes file rewriting thus forcing rebuilding project (for languages like C++, etc).

SmartGit has option "+keep working copy" in stash dialog.

### Environment

- Git Extensions 4.0.2.16100
- Build 25100ec1f7c8da8f798613da74826038af81a50e
- Git 2.40.0.windows.1
- Microsoft Windows NT 10.0.19043.0
- .NET 6.0.16
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 3.1.24 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by mstv** [comment]
Are you looking for "Stash staged changes"?

![image](https://user-images.githubusercontent.com/36601201/233781995-8e26b143-15ef-4e38-b41b-f68e7a0f69ba.png)

![image](https://user-images.githubusercontent.com/36601201/233782032-0cbf109f-eee1-4d32-9abd-c7fd1d8a712a.png)

**Comment by vovodroid** [comment]
> Are you looking for "Stash staged changes"?

No. This command reset stage and local files. There is option stash with "keep index", but it overwrites local file.

**Comment by RussKie** [comment]
Why not just commit this changes?

**Comment by vovodroid** [comment]
> Why not just commit this changes?

It's not the same as create local backup for upgoing work. For instance you may save code before some refactoring. Commit will hide local changes.

**Comment by RussKie** [comment]
> > Why not just commit this changes?
> 
> It's not the same as create local backup for upgoing work. For instance you may save code before some refactoring. Commit will hide local changes.

Not sure that I follow. I think I can count the number of times I used stash in the past decade on one hand... Could you please help me understand with a concrete scenario?

Also, say, you convince us the app should have the feature you ask for. How would you achieve the desired from a command line?

**Comment by vovodroid** [comment]
> Could you please help me understand with a concrete scenario?

Assume that during development I would like to make many further changes, but I also want to save current changed code. 

> How would you achieve the desired from a command line?

I'm almost not familiar with git command line, but SmartGit does the following:

```
git.exe  stash push --include-untracked --message kkk -- project\settings.py
git.exe stash apply --index stash@{0}

```
It seems SG save and then reapply stash. May be I'm wrong about not touching file time, but anyway it could be handy to have option to stash and keep working copy:

![image](https://user-images.githubusercontent.com/8691781/233831022-6660f682-568d-4ca8-82c2-32b9846df412.png)

**Comment by RussKie** [comment]
> Assume that during development I would like to make many further changes, but I also want to save current changed code.

Isn't what you achieve by making a commit?

**Comment by vovodroid** [comment]
Commit hides all local changes. We are talking about development process, before it's ready to commit.

**Comment by RussKie** [comment]
Git is about commits. There's no restrictions on how many commits a
developer can or should have.
This allows iterating over implementations and exploring different
hypothesise. Once the desired result is achieved - perform an interactive
rebase reverting unwanted changes, squashing, reordering, etc. those
changes that still applicable....
Many commits also allow you to compare and review different point in time.

**Comment by vbjay** [comment]
I'm with @RussKie on this. I have created a feature branch and bounced ideas and made wip commits so many times. Played with an idea and said nope that doesn't work and reset to an earlier commit. However I still had reflog to oh I wrote this code before and can copy code or reset file to x commit version in reflog. If I need to squash wip commits I can before mergunt into a trunk branch. You can always compare commits and even reset a file or all files to that commit in GE. I think the reflog and commits beat stash. 

Where I use stash is if I have made changes and need to switch to another branch because of a found bug or other reason or a rebase while I have changes using autostash. Otherwise I use commits instead.

**Comment by pmiossec** [comment]
I agree that it's maybe not a feature that we use every day so maybe not
having it in direct access in the toolbar.

But in the other hand, the feature exists in the command line and a user
find it convenient so I don't think we should refuse to add it if a pull
request is done.

As a consequence, for me, the best place to have the feature is maybe as a
checkbox in the manage stash form where we can create a stash...

**Comment by pmiossec** [comment]
Thinking about it again, a split button for the 'Create stash' instead of a
checkbox could be a better idea...

**Comment by vovodroid** [comment]
> maybe as a checkbox in the manage stash form where we can create a stash...

Exactly:
![image](https://user-images.githubusercontent.com/8691781/233837609-4602c45b-4aa6-4832-a60c-11a3c4267b15.png)

**Comment by vovodroid** [comment]
> a split button for the 'Create stash' instead of a checkbox could be a better idea...

I guess check box is better, as stash option like "Keep index".

**Comment by RussKie** [comment]
@vovodroid if you know the git command, and it is something you need now - you can add a custom script (see Settings > Scripts).

**Comment by vovodroid** [comment]
But "Keep working tree" is not something that should be always done.

**Comment by gerhardol** [comment]
I use stash often for temporary changes that are not something worth creating a branch for, just some I keep in the back of my mind for some time. Minor need to clean stashes.
A stash is almost always not ready for a commit, some changes should be discarded.

https://git-scm.com/docs/git-stash

There is no direct option, it has to be implemented with stash-push followed by stash-apply.
So the modification time is changed.

Personally, I would just stash and restore the changes I would like to keep in scenarios like this.
A "keep worktree" is a minor deviation from Git behavior.

I would not vote against a PR with this change (that also describes the behavior in the documentation), but it is not something I would spend time on.
A script is a better solution - such scripts could be included by default too.
